### PR TITLE
Benchmark tool for DIFI Source block

### PR DESCRIPTION
**Pull request by 777arc**
Created a benchmark test for the DIFI Source block, so that one can measure the performance of the block before and after future optimization occurs.  The intent is to characterize what is the maximum data rate supported by the DIFI Source block running within GNU Radio.  We want to be able to, for example, tweak GNU Radio runtime and buffer params and see if there's an improvement, and ensure functionality.

Currently the output is a plot of % packets received over sample rate (see example in ADO) which seemed more useful than just a single number, example output:

benchmark_difi_source_example

[AB#10918](https://ag-ascii.visualstudio.com/a2b18d2f-d51c-4194-97e8-d6244d4a0156/_workitems/edit/10918)

### CODE REVIEW

File: gr-azure-software-radio/utils/benchmark_difi_source.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        tb = DifiSourceFlowgraph()
+
+        # Send DIFI packets
+        print("Starting send, delay =", delay)
+        tb.start()  # At this point the flowgraph is running in C++ in the background (the python below is feeding packets but its not going to slow it down)
+        start_t = time.time()
+        num_repeats = 40  # manually tweaked this number to give a tight curve while as low as possible
+        for ii in range(num_repeats):
+            for msg in msgs:
+                sock_tx.send(msg)
```
Comment (jo-rivera) [inline gr-azure-software-radio/utils/benchmark_difi_source.py:86]: should we consider using iperf to leverage other metrics as well? https://iperf3-python.readthedocs.io/en/latest/examples.html

**Review by jo-rivera — CHANGES_REQUESTED** [review]

File: gr-azure-software-radio/utils/benchmark_difi_source.py (hunk 1 above)
Comment (jo-rivera) [reply gr-azure-software-radio/utils/benchmark_difi_source.py:86]: never mind... we are testing in loopback. It makes sense when we have the client and server in different machines, and we are sending traffic over the network

**Review by jo-rivera — CHANGES_REQUESTED** [review]

**Comment by jo-rivera** [comment]
can we include instructions on how to run the benchmark test? does the user just run the flowgraph? what environment was used to run the test and generate the attached plot? (VM?)

File: gr-azure-software-radio/utils/benchmark_difi_source.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        self.difi_sink = azure_software_radio.difi_sink_cpp_fc32(0, 0, '127.0.0.1', 3333, bool(
+            False), samples_per_packet, 0, int(352), int(samp_rate), int(0), 0, 1000, int(72), int(8), int(2), 127, 0, 1.0, -1.0)
+        self.sig_source = analog.sig_source_c(samp_rate, analog.GR_SIN_WAVE, 1000, 0.5, 0, 0)
+        self.throttle = blocks.throttle(gr.sizeof_gr_complex*1, samp_rate, True)
+        self.connect(self.sig_source, self.throttle)
+        self.connect(self.throttle, self.difi_sink)
+
+tb = DifiSinkFlowgraph()
+
+# Generate and save a bunch of DIFI packets, we start the flowgraph then start recv()'ing
```
[the remark this answers is not in the document]
Comment (777arc) [reply gr-azure-software-radio/utils/benchmark_difi_source.py:41]: Rather keep it all in 1 file for convenience since it's a utility

**Comment by 777arc** [comment]
> can we include instructions on how to run the benchmark test? does the user just run the flowgraph? what environment was used to run the test and generate the attached plot? (VM?)

I'll stick the details in the png itself

**Review by jo-rivera — APPROVED** [review]

**Pull request merged**